test/pthread/test_pthread_kill.c

```diff
@@ -13,52 +13,58 @@
 #include <signal.h>
 
 _Atomic int sharedVar = 0;
+_Atomic int gotTermSignal = 0;
 
-static void *thread_start(void *arg) {
-  // As long as this thread is running, keep the shared variable latched to nonzero value.
-  for (;;) {
-    ++sharedVar;
+pthread_t thr;
+
+void signal_handler(int sig, siginfo_t * info, void * arg) {
+  printf("signal: %d onthread=%d\n", sig, pthread_self() == thr);
+  if (sig == SIGTERM) {
+    gotTermSignal = 1;
   }
+}
 
-  pthread_exit(0);
+void setup_handler() {
+  struct sigaction act;
+  sigemptyset(&act.sa_mask);
+  act.sa_flags = SA_SIGINFO;
+  act.sa_sigaction = signal_handler;
+  sigaction(SIGTERM, &act, NULL);
 }
 
-void BusySleep(long msecs) {
+
+void sleepms(long msecs) {
   usleep(msecs * 1000);
 }
 
+void *thread_start(void *arg) {
+  // As long as this thread is running, keep the shared variable latched to nonzero value.
+  while (!gotTermSignal) {
+    sleepms(1);
+    ++sharedVar;
+  }
+  printf("got term signal, shutting down thread\n");
+  pthread_exit(0);
+}
+
 int main() {
-  pthread_t thr;
+  setup_handler();
 
   sharedVar = 0;
   int s = pthread_create(&thr, NULL, thread_start, 0);
   assert(s == 0);
 
   // Wait until thread kicks in and sets the shared variable.
-  while (sharedVar == 0)
-    BusySleep(10);
-  printf("thread startd\n");
-
-  s = pthread_kill(thr, SIGKILL);
-  printf("thread killed\n");
-  assert(s == 0);
-
-  // Wait until we see the shared variable stop incrementing. (This is a bit heuristic and hacky)
-  for(;;) {
-    int val = sharedVar;
-    BusySleep(100);
-    int val2 = sharedVar;
-    if (val == val2) break;
+  while (sharedVar == 0) {
+    sleepms(10);
```

Comment:
Can we use condition variables instead of sleeps to make the test run as fast as possible?

Reply:
Done: https://github.com/emscripten-core/emscripten/pull/22479